### PR TITLE
How to Unbundle Nginx from Omnibus GitLab for Serving Multiple Websites

### PR DESCRIPTION
**Pull request by programarivm**
Hello there,
It's Jordi, a new linode contributor.
I just optimized one of my clients' servers and wrote this tutorial.
Thank you.

### CODE REVIEW

**Comment by pbzona** [comment]
Hi @programarivm 

We're focusing most of our new Ubuntu guides on 16.04, have you tested this guide on that version? What other types of sites/applications were the virtual hosts intended to handle? I have to wonder whether it's wise to host a live site, for example, on the same Linode as GitLab. 

This seems like an interesting topic, but my main concern is how it would be applied in practice.

**Comment by programarivm** [comment]
Hi pbzona,

I'm glad to hear from someone at Linode.

I am sorry to tell you that I'm not sure about your concern at this moment, I simply optimized an Ubuntu 14.04 LTS machine and shared that how-to.

Regards,

**Comment by pbzona** [comment]
Jordi,

I guess my question was more about what you mean by "optimize." When you unbundle Nginx from GitLab, the purpose seems to be to make its virtual hosting available for other purposes (aside from GitLab). How would one use the extra virtual hosting in practice? What was the ultimate goal in optimizing your client's machine? What were they trying to achieve?

**Comment by programarivm** [comment]
Hello,

They wanted to use Nginx's virtual hosting capability.

The problem with having Nginx bundled to GitLab is that you only can install one web app on one server.

So, this how-to helps you turn your GitLab's Nginx into a multi-purpose web server for you to install as many virtual servers as you want.

"Saving money" is a funny way to say what this how-to is intended to be about.

Regards,

**Comment by programarivm** [comment]
Hello, sure it isn't wise at all to have a live site on the same Linode as GitLab. They wanted to host a number of admin web apps on that server.

**Comment by pbzona** [comment]
Thanks for the clarification, that makes a lot more sense. I'm going to do some additional research to see if this is something we'd be interested in publishing. We'll be in touch with an answer soon.

**Comment by programarivm** [comment]
I'm happy to hear that. If that is the case feel free to adapt it to your own writing style and audience.

**Comment by pbzona** [comment]
@programarivm We're interested in the topic and would like to work with you to develop this into something we can publish. Over the next few days I'll do some testing to make sure everything works as you've described it, and possibly making some additions/changes. I'll follow up with a pull request once I've gotten through everything. Thanks for your patience.

**Comment by programarivm** [comment]
Thank you pbzona. Regards,

**Comment by pbzona** [comment]
@programarivm The guide seems to be focused on upgrading GitLab from an existing, previous version. However, I don't have a previous instance, so I'm installing it on a fresh server. When I run your commands and set up the vhost, all I'm able to get is the default nginx welcome page on my domain (I set it up for HTTP rather than HTTPS as described in your vhost, with the same result). 

Was there any additional configuration required prior to these steps? Could you tell me a little bit more about the environment you tested this in - specifically how you installed and configured GitLab and nginx initially?

**Comment by programarivm** [comment]
@pbzona You're right. I wasn't specific enough, but I just updated the document in order to fix that. Could you please read it again? Thank you.

**Comment by pbzona** [comment]
@programarivm Thanks for your update. I'm still running into two problems. When I try to open my domain in a browser after following the guide:

1) It tries to force an HTTPS connection to `https://xxxxxxxx.net/users/sign_in`. Since I didn't configure SSL, this won't work and the connection fails.
2) When I remove the S, and try to connect to`http://xxxxxxxx.net/users/sign_in`, a 404 is returned.

Since the example vhost file you provided includes SSL directives, I thought maybe that's where the error was. I installed GitLab normally and grabbed the bundled nginx's vhost file (or what would be equivalent to it) from `/var/opt/gitlab/nginx/conf/gitlab-http.conf`. This is included in the bundled nginx.conf file. Here are the contents of that file: 

https://gist.github.com/pbzona/93841c459008c53a95a8b825418c49ed

I tried that in the unbundled version's vhost file and still no luck. I had thought the issue was with virtual hosting, but now I'm not so sure. When you set yours up, was there any additional configuration you did to make nginx work?

**Comment by pbzona** [comment]
@programarivm Could you please hold off on additional commits (unless technical) until I make a pull request? I appreciate your help, but it's going to cause a merge conflict when I finish my review and push it.

**Comment by programarivm** [comment]
@pbzona I'm so sorry for that. I just made a technical commit with significant changes. The document was updated for the newest version of GitLab CE Omnibus.

**Comment by programarivm** [comment]
@pbzona That's right. Let's make things simple. The SSL thing has finally been removed because that is not the main subject of this guide. Thank you for your feedback.

**Comment by pbzona** [comment]
@programarivm No problem at all. It looks like the guide was missing some dependencies (which are included in the link you provided), and everything seems to be working. I'll be doing some additional testing and editing over the next day or two, and should have a pull request to you by then. Thanks for your help!

**Comment by pbzona** [comment]
Hi @programarivm, I've made my pull request here: https://github.com/programarivm/docs/pull/1

Please let me know if you have questions!

File: docs/applications/development/how-to-unbundle-nginx-from-omnibus-gitlab-for-serving-multiple-websites.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+## Create a New Virtual Host
+
+In this section, we'll create a new virtual host to serve GitLab. Since we unbundled nginx, we'll also be able to configure other virtual hosts for other websites and apps.
+
+1.  Copy the default virtual host file to a new virtual host file, replacing `example.com` with the name you'd like to use for your virtual host:
+
+	    sudo cp /etc/nginx/sites-available/default /etc/nginx/sites-available/example.com
+
+2.  Edit your new virtual host file to match the following, replacing `example.com` with your own hostname:
```
Comment (alexfornuto) [inline docs/applications/development/how-to-unbundle-nginx-from-omnibus-gitlab-for-serving-multiple-websites.md]: If the file excerpt below has the only directives we need, why do we copy the default host first?

**Comment by programarivm** [comment]
Hello @pbzona and @alexfornuto. We copy the default virtual host file to a new file so that we can benefit from having a template file ready to be reused. Let's say that is the rule of thumb — copy from default virtual config — to be considered when creating new virtual hosts.

**Comment by programarivm** [comment]
Please be so kind to add the phrase above if it turned out that I wasn't specific enough. Thank you so much. Regards.

**Comment by alexfornuto** [comment]
@programarivm I performed a tech test, but only made minor edits for copy and formatting. https://github.com/programarivm/docs/pull/2

Regarding my note on Line 146, the default virtual host has to much more that has to be culled that I don't feel it's worth it. The line should be adjusted to instruct the reader to make a new file and copy the file you provided, changing the appropriate values.

**Comment by programarivm** [comment]
Thank you @alexfornuto. I see your point, in which case this is a minor issue in my opinion. Please feel free to proceed as you like. Best regards.

**Comment by alexfornuto** [comment]
@programarivm we're almost ready to publish this guide! I need to know if you prefer your bounty as Linode credit or through PayPal. In either case, what email address should we use?

**Comment by programarivm** [comment]
I'm happy to hear that! It was a pleasure working with you. I just sent the information to the email above. Thank you.

**Comment by EdwardAngert** [comment]
Hi @programarivm,
We received your email and have submitted a payment request to our processing department. The payment should clear in the next few business days.
This branch doesn't show as merged because we squashed a few of your proofreading commits into one, but the commit history in the master repo still shows your contribution.

Thanks again,
Edward

**Pull request closed without merging**